tests/host_callback_to_tf_test.py

```diff
@@ -0,0 +1,256 @@
+
+  # Define the fwd and bwd custom_vjp functions
+  def make_call_vjp_fwd(arg):
+    return make_call(arg), arg  # Return the primal argument as the residual
+
+  def make_call_vjp_bwd(res, ct_res):
+    arg = res  # residual is the primal argument
+
+    def tf_vjp_fun(arg_and_ct_res):
+      """Invoke TF gradient; used with hcb.call."""
+      arg, ct_res = arg_and_ct_res
+
+      def make_var(a):
+        return a if isinstance(a, tf.Variable) else tf.Variable(a)
+
+      arg_var = tf.nest.map_structure(make_var, arg)
+
+      with tf.GradientTape(persistent=True) as tape:
+        res = tf_fun(arg_var)
+
+      tf.nest.assert_same_structure(res, ct_res)
+      accumulator = None  # Accumulate argument cotangent. Same structure as "arg"
+
+      def acc_ct(res_, ct_res_):
+        dres_darg = tape.gradient(res_, sources=arg_var,
+                                  unconnected_gradients=tf.UnconnectedGradients.ZERO)
+        tf.nest.assert_same_structure(dres_darg, arg)
+        scaled_dres_darg = tf.nest.map_structure(lambda d: d * ct_res_, dres_darg)
+        nonlocal accumulator
+        accumulator = (scaled_dres_darg if accumulator is None
+                       else tf.nest.map_structure(lambda x, y: x + y,
+                                                  accumulator, scaled_dres_darg))
+
+      tf.nest.map_structure(acc_ct, res, ct_res)
+      return accumulator
+
+    return (call_tf_full_ad(tf_vjp_fun, (arg, ct_res),
+                            result_shape=arg),)
+
+  make_call.defvjp(make_call_vjp_fwd, make_call_vjp_bwd)
+  return make_call(arg)
+
+
+CALL_TF_IMPLEMENTATIONS = {
+    "none": call_tf_no_ad,
+    "simple": call_tf_simple_ad,
+    "full": call_tf_full_ad,
+}
+
+
+class CallToTFTest(jtu.JaxTestCase):
+
+  def setUp(self):
+    if tf is None:
+      raise unittest.SkipTest("Test requires tensorflow")
+
+  @parameterized.named_parameters(
+      dict(
+          testcase_name=f"_ad={ad}",
+          ad=ad)
+      for ad in CALL_TF_IMPLEMENTATIONS.keys())
+  def test_impl(self, ad="simple"):
+    call_tf = CALL_TF_IMPLEMENTATIONS[ad]
+
+    def f_jax(x):
+      return jnp.sin(x)
+
+    def f_outside(x):
+      return call_tf(tf.math.sin, x,
+                     result_shape=x)
+
+    res = f_outside(3.)
+    self.assertAllClose(f_jax(3.), res)
+    self.assertAllClose(f_jax(3.), api.jit(f_outside)(3.))
+
+  @parameterized.named_parameters(
+      dict(
+          testcase_name=f"_ad={ad}",
+          ad=ad)
+      for ad in CALL_TF_IMPLEMENTATIONS.keys()
+      if ad != "none")
+  def test_grad(self, ad="simple"):
+    call_tf = CALL_TF_IMPLEMENTATIONS[ad]
+
+    def f_jax(x):
+      return 3. * jnp.sin(2. * x)
+
+    def f_outside(x):
+      return 3. * call_tf(tf.math.sin, 2. * x, result_shape=x)
+
+    x = 4.
+    self.assertAllClose(f_jax(x), f_outside(x))
+
+    # print(f"outside_jaxpr = {api.make_jaxpr(f_outside)(xy)}")
+    # print(api.xla_computation(f_outside)(4.).as_hlo_text())
```

Comment:
delete? or add to the message field of `assertAllClose`?

Reply:
Done